index.js

```diff
@@ -3,36 +3,40 @@ const { derivePath } = require('near-hd-key');
 const bs58 = require('bs58');
 const nacl = require('tweetnacl');
 
-const KEY_DERIVATION_PATH = "m/44'/397'/0'"
+const KEY_DERIVATION_PATH = "m/44'/397'/0'";
 
 const generateSeedPhrase = (entropy) => {
-    return parseSeedPhrase(entropy !== undefined ? bip39.entropyToMnemonic(entropy) : bip39.generateMnemonic())
-}
+    return parseSeedPhrase(entropy !== undefined ?
+        bip39.entropyToMnemonic(entropy) : bip39.generateMnemonic());
+};
 
-const normalizeSeedPhrase = (seedPhrase) => seedPhrase.trim().split(/\s+/).map(part => part.toLowerCase()).join(' ')
+const normalizeSeedPhrase = (seedPhrase) => seedPhrase.trim()
+    .split(/\s+/)
+    .map(part => part.toLowerCase())
+    .join(' ');
 
 const parseSeedPhrase = (seedPhrase, derivationPath) => {
-    const seed = bip39.mnemonicToSeed(normalizeSeedPhrase(seedPhrase))
-    const { key } = derivePath(derivationPath || KEY_DERIVATION_PATH, seed.toString('hex'))
-    const keyPair = nacl.sign.keyPair.fromSeed(key)
-    const publicKey = 'ed25519:' + bs58.encode(Buffer.from(keyPair.publicKey))
-    const secretKey = 'ed25519:' + bs58.encode(Buffer.from(keyPair.secretKey))
-    return { seedPhrase, secretKey, publicKey }
-}
+    const seed = bip39.mnemonicToSeed(normalizeSeedPhrase(seedPhrase));
+    const { key } = derivePath(derivationPath || KEY_DERIVATION_PATH, seed.toString('hex'));
+    const keyPair = nacl.sign.keyPair.fromSeed(key);
+    const publicKey = 'ed25519:' + bs58.encode(Buffer.from(keyPair.publicKey));
+    const secretKey = 'ed25519:' + bs58.encode(Buffer.from(keyPair.secretKey));
+    return { seedPhrase, secretKey, publicKey, keyPair };
```

Comment:
Hm, I think, ideally, this should be a class instance. Much like in [`near-api-js`](https://github.com/near/near-api-js/blob/37ca2222e7afda3b77929d695b9d84bba21b418e/packages/near-api-js/src/utils/key_pair.ts#L56).

```js
class PublicKey extends Uint8Array {
  toString(): string;
}

class SecretKey extends Uint8Array {
  toString(): string;
}

class SeedPhrase extends string {
  publicKey: PublicKey;
  secretKey: SecretKey;
}
```

What do you think @frol?

Reply:
Another approach is to default all of them to `string`s, retaining current behavior, and including the raw buffer in a field.

```js
class PublicKey extends string {
  data: Uint8Array;
}

class SecretKey extends string {
  data: Uint8Array;
}

class SeedPhrase extends string {
  publicKey: PublicKey;
  secretKey: SecretKey;
}
```

Reply:
Well, classes are not widely used in JS world, so returning an object with bare fields is a common approach. I don't mind that approach, but I don't like having two sources of truth (string and bytes are disconnected), and only the first proposed approach (in the first message of this thread) meets that requirement, but it will introduce a breaking change.

I would try to avoid introducing breaking changes unless we plan to redesign this small package.

This gets us back to the square of one.

@miraclx I don't want to be the bottleneck for the tooling team, and I have never maintained this project, so feel free to take it over and proceed without my reviews.

Reply:
Yeah, In that case, we should go with the path of least resistance, and reduce the scope of impact for this change, so it's purely additive. @loosefer, can you try implementing the second approach?

Reply:
@miraclx Unfortunatetly, no possible ways to implement your suggestion. I like your idea and trying to research how we can extend prototype of String without of changing global prototype ([because it is anti-pattern](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/proto)), but any implementation have to call methods valueOf or toString() that broke the backward compatibility.

For example:
![telegram-cloud-photo-size-2-5332672914618695689-x](https://user-images.githubusercontent.com/4532214/194540982-60c35163-2a2e-4711-94f5-4c0e4deb4c8e.jpg)

- it extends prototype of `String` globally and makes property static

Next way is create instance of String, but invoking of variable returns instance of String. toString() should be called
<img width="255" alt="image" src="https://user-images.githubusercontent.com/4532214/194539253-19a3927a-016b-4751-9170-36354f0eb0a5.png">

One more attempt is use[inherit function for prototype nesting for Google Closure Library](https://github.com/google/closure-library/blob/master/closure/goog/base.js#L1847) and modify it, but it the same as `extends` syntax sugar in ES6 and working with `objects`, but we need a `string`.

So, i totally agree with @frol about two sources of truth, but can't extend string type as you suggest.
Let's freeze this pull requests about 2-3 days and i try to find best solution during weekend =)

Reply:
This should work:

```js
class PublicKey extends String {
    constructor(key) {
        super('ed25519:' + bs58.encode(Buffer.from(key)));
        this.data = key;

        Object.freeze(this); // <- optional
    }
}
```

Reply:
@miraclx 
Seems like It can broke the backward compatibility.
Before, all interfaces using [secretKey and publicKey](https://github.com/near/near-seed-phrase/blob/master/index.js#L20) as `string` type. After refactor, type of key will be `object`. Sure, it can be dynamic casted to `string`, but it will be implicit. In other words, `typeof secretKey === 'string'` will be changed to `typeof secretKey === 'object'`.

I'm afraid it can create a lot of problems when version of library be bumped for users and want to be very carefully with compatibility. Thank you.